File: src/a2/VirtualFriends.java
package a2;

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class VirtualFriends {

    private Map <String, Node> map = new HashMap<>();

            class Node {
                String name;
                Node parent;
                int noOfFriends;

            }
            /*
                create a set with one user
             */
            public void makeSet(String name){
                if(!map.containsKey(name)) {

                    Node node = new Node();
                    node.name = name;
                    node.parent = node;
                    node.noOfFriends = 1;

                    map.put(name, node);
                }
            }

//            public String findSet(String name){
//                return findSet(map.get(name).name);
//            }

            /*
                Find set representative recursively
             */
            private Node findSet(Node node){
                Node parent = node.parent;
                if(parent == node){
                    return parent;
                }
                node.parent = findSet(node.parent);
                return node.parent;

            }

            /*
                Combines two sets together
                @return true if successful and false if they're in the same set
             */

            public boolean union(String user1, String user2){
                Node nodeA = map.get(user1);
                Node nodeB = map.get(user2);

                Node parentA = findSet(nodeA); //return representative
                Node parentB = findSet(nodeB);

                if(parentA.name.equals(parentB.name)) return false;

                if(parentA.noOfFriends > parentB.noOfFriends){
                    parentA.noOfFriends = (parentA.noOfFriends + parentB.noOfFriends);
                    parentB.parent = parentA;
                } else {
                    parentB.noOfFriends = (parentB.noOfFriends + parentA.noOfFriends);
                    parentA.parent = parentB;
                }

                return true;

            }

            public static void main (String [] args){

                VirtualFriends test = new VirtualFriends();

                Scanner input = new Scanner(System.in);
               //System.out.println("Enter No of Test Cases");

                int testCases = Integer.parseInt(input.nextLine());

               // System.out.println("FriendShips");
                int friendShips = Integer.parseInt(input.nextLine());

                while(testCases>0){


                    while(friendShips > 0){
                        //System.out.println("Make friends");
                        String user1 = input.next();
                        String user2 = input.next();


                        test.makeSet(user1);
                        test.makeSet(user2);
                        if(test.union(user1,user2)){
                            System.out.print(test.findSet(test.map.get(user1)).noOfFriends);
                        } else {
                            System.out.print(test.findSet(test.map.get(user1)).noOfFriends);
                        }



                        //if(input.hasNext()) input.nextLine();

                        friendShips--;
                    }
                    testCases--;
                }




            }





}

File: src/a2/virtualfriends.cpp
#include<bits/stdc++.h> 
using namespace std; 
#define MAXSIZE 100000 

vector<int>G[MAXSIZE]; 

int parent[MAXSIZE]; 
int noOfFriends[MAXSIZE];
int _rank[MAXSIZE]; 

//recursively find representative of the set
int _findSet(int elem){ 
if(elem==parent[elem]) return elem;
     
      return parent[elem]=_findSet(parent[elem]);
} 

int _union (int user1,int user2) { 
    user1 = _findSet(user1); 
    user2 = _findSet(user2); 

    int ans=0; 

    if(user1 != user2){ 
        
        if( _rank[user1] > _rank[user2] ){ 
            parent[user2]=user1; 
            noOfFriends[user1]+=noOfFriends[user2]; 
            
        } else { 
            parent[user1]=user2; 
            noOfFriends[user2]+=noOfFriends[user1];
            
        } 
        
        if( _rank[user2] == _rank[user2] ){ 
            ++_rank[user1]; 
            noOfFriends[user1]++; 
   
        } 

    } 

    return max(noOfFriends[user1],noOfFriends[user2]); 

} 

int main() { 
    int testCases; 

    scanf("%d",&testCases); 

    while(testCases--) { 
        int friendShips; 
        scanf("%d",&friendShips); 
        int count=0; 
        map<string,int>myMap; 

    for(int i=0;i<=friendShips;i++)parent[i]=i,_rank[i]=0,noOfFriends[i]=1; 
    
    for(int i=0;i<friendShips;i++){ 

    char user1[25],user2[25]; 

    scanf("%s",user1);
    scanf("%s",user2); 

    if(!myMap[user1])myMap[user1]=++count; 

    if(!myMap[user2])myMap[user2]=++count; 

    int friends= _union(myMap[user1],myMap[user2]); 
    printf("%d\n",friends); 
    }

    } 
return 0; 

}



